docs/source/upgrade.rst

```diff
@@ -1,819 +1,124 @@
+   v2 = chainer.Variable(cupy.arange(10, dtype=cupy.float32))  # GPU
 
-.. admonition:: Example
+   # The line below raises an exception, because arguments are on different device.
+   F.maximum(v1, v2)
 
-   Consider the following model definition and the code to call it in test mode written for Chainer v1.
+Prior to v4, the above code raises an exception like ``ValueError: object __array__ method not producing an array``, which was difficult to understand.
+In v4, the error message would become ``ValueError: numpy and cupy arrays are mixed in the forward input (Maximum)``.
+This kind of error usually occur by mistake (for example, not performing ``to_gpu`` for some variables).
 
-   .. code-block:: py
+.. attention::
 
-      # Chainer v1
-      import chainer.functions as F
+   As the argument validation is strictened, call of new-style functions (inheriting :class:`FunctionNode` class) intentionally mixing NumPy/CuPy arrays in arguments will not work in Chainer v4.
+   Please transfer all arrays to the same device before calling functions.
 
-      class MyModel(chainer.Link):
-          ...
+Note that old-style function classes (inheriting :class:`Function`) are not affected.
 
-          def __call__(self, x, train=True):
-              return f(F.dropout(x, train=train))
+Reproducibility Improvement with MultiProcessIterator
+-----------------------------------------------------
 
-      m = MyModel(...)
-      y = m(x, train=False)
+Prior to v4, NumPy random states are shared between threads when using :class:`chainer.iterators.MultiprocessIterator`.
+This release fixed this issue by using a separate random state for a sub thread to improve reproducibility of experiments.
+See the discussion in `#3575 <https://github.com/chainer/chainer/pull/3575>`_ for more details.
 
-   In Chainer v2, it should be updated into the following code:
+Update of Docker Images
+-----------------------
 
-   .. code-block:: py
+Chainer official Docker images (see :doc:`install` for details) are now updated to use CUDA 8.0 and cuDNN 6.0.
+This change was introduced because CUDA 7.5 does not support NVIDIA Pascal GPUs.
 
-      # Chainer v2
-      import chainer.functions as F
 
-      class MyModel(chainer.Link):
-          ...
+Chainer v3
+==========
 
-          def __call__(self, x):
-              return f(F.dropout(x))
+Introduction of New-style Functions
+-----------------------------------
 
-      m = MyModel(...)
-      with chainer.using_config('train', False):
-          y = m(x)
+This release introduces new-style functions (classes inheriting from :class:`FunctionNode`) that support double backward (gradient of gradient).
+See the `Release Note for v3.0.0 <https://github.com/chainer/chainer/releases/tag/v3.0.0>`_ for the usage of this feature.
 
-.. _upgrade-configurations:
+Many of :doc:`reference/functions` are already migrated to new-style, although some of functions are still old-style (classes inheriting from :class:`Function`).
```

Comment:
Do we need this paragraph?
Instead, we should state that writing new functions with `FunctionNode` is encouraged.

Reply:
In this paragraph I mean "currently not all functions supports double backprop". I'll make it clear.

> we should state that writing new functions with FunctionNode is encouraged.

I agree with this.